### PR TITLE
WIP: .NET Conventions Refactoring

### PR DESCRIPTION
**Pull request by EslaMx7**
> **AI Generated Commit Message: Refactor tokenizers and enhance vocab handling**
This commit significantly refactors the BaseTokenizer and related classes, enhancing their modularity, interface simplicity, and internal capabilities. Key changes include the removal of outdated porting comments, a shift in method return types for better interface clarity, and the elimination of unnecessary methods. Notably, the Tokenize method now returns a list instead of a vector, and several methods have been made private to tighten the class's public interface.
The design of XLMRobertaTokenizer has been refined, with fields now set as readonly to enforce immutability, and redundant functionality has been removed or integrated into BaseTokenizer. Corresponding tests have been adjusted to reflect these changes.
Vocabulary handling has been significantly improved across interfaces, with new methods added to support a broader range of special tokens directly. This standardizes and simplifies the management of special tokens across different tokenizer implementations.
Overall, these changes aim to improve the tokenizer framework's design, efficiency, and ease of use, making it a more powerful tool for text processing and analysis.

## Work In Progress:

- [ ] Using `byte[]` over `string`
- [ ] Using custom types over tuples
- [ ] Rename functions/variables to follow dotnet conventions
- [ ] Update Readme file
- [ ] Nuget Package

### CODE REVIEW

File: src/Tokenizer/BaseTokenizer.cs [hunk 1]
```diff
@@ -296,6 +286,61 @@ public List<long> ConvertTokensToIds(List<string> tokens)
         };
     }
 
+    /// <summary>
+    /// Builds input with special tokens.
+    /// </summary>
+    public TokenIdsWithSpecialTokens BuildInputWithSpecialTokens(TokenIdsWithOffsets tokens1, TokenIdsWithOffsets tokens2 = null)
```
Comment (ClementLokad) [inline src/Tokenizer/BaseTokenizer.cs]: We want to remove the `XLMRobertaTokenizer tokenizer` parameter from `encode` but we want to keep the possibility to implement another tokenizer in the future. A tokenizer can have a custom method `BuildInputWithSpecialTokens` as it is the case for the `XLMRobertaTokenizer`.
I would recommend having a virtual `BuildInputWithSpecialTokens` in `BaseTokenizer` [with this implementation](https://github.com/guillaume-be/rust-tokenizers/blob/main/main/src/tokenizer/base_tokenizer.rs#L1168) and keeping this one in `XLMRobertaTokenizer`  and adding the `override` keyword to the method, just like for `TokenizeToTokens`.

File: src/Vocab/IVocab.cs [hunk 2]
```diff
@@ -2,14 +2,22 @@
...
+
     Dictionary<string, long> Values { get; }
     Dictionary<long, string> Indices { get; }
     Dictionary<string, long> SpecialValues { get; }
     Dictionary<long, string> SpecialIndices { get; }
 
     IEnumerable<string> SpecialTokens();
 
+    string GetUnknownValue();
+    string GetBosValue();
```
Comment (ClementLokad) [inline src/Vocab/IVocab.cs:14]: We can remove these methods if it's not needed anymore after the other comment.

**Review by ClementLokad — CHANGES_REQUESTED** [review]

File: src/Tokenizer/BaseTokenizer.cs (hunk 1 above)
Comment (EslaMx7) [reply src/Tokenizer/BaseTokenizer.cs]: Agree 100%
Done in next commit.

File: src/Vocab/IVocab.cs (hunk 2 above)
Comment (EslaMx7) [reply src/Vocab/IVocab.cs:14]: They are still in use for a few test cases; I will ensure to remove them later if they become unnecessary.

**Pull request merged**